kvcoord: extend txn refresh error with the failed keys

lidorcarmel opened the pull request:
- kvcoord: use a structured error for txn refresh error

  Return a structured error instead of a string. Later we will use
  the error to print better error message, containing the specific
  key that we failed refreshing. This is step 1 out of 2 to
  fix issue #61838.

  Release note: None

- kvcoord: extend txn refresh error with the failed keys

  Use the new structured error to log the actual key that we failed
  refreshing. Instead of:
  .. failed preemptive refresh ..

  The new error will print:
  .. failed preemptive refresh due to a conflict: committed value on key <key>
  .. failed preemptive refresh due to a conflict: intent on key <key>

  Rename tryUpdatingTxnSpans() to tryRefreshTxnSpans().

  Fixes #61838.

  Release note: None

Commit:
kvcoord: use a structured error for txn refresh error

Return a structured error instead of a string. Later we will use
the error to print better error message, containing the specific
key that we failed refreshing. This is step 1 out of 2 to
fix issue #61838.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73413)
<!-- Reviewable:end -->

erikgrinaker reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-Mpzg17pCOn-SeC0FlGF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lidorcarmel)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MpzfjJ-EASsgAC2YW6Q:-MpzfjJ-EASsgAC2YW6R:bhk2k78) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e56b362a009fae06d3be8567f75fe36e39330d/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L492)):*
> ```Go
> 	require.Nil(t, br)
> 	require.NotNil(t, pErr)
> 	require.Regexp(t, `TransactionRetryError: retry txn \(RETRY_SERIALIZABLE - failed preemptive refresh of span\(s\): \[\{a\-b\}\]\)`, pErr)
> ```

Consider `require.Contains` instead, to avoid the regexp escapes.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MpzfjJ-EASsgAC2YW6Q:-Mq0SvfVDfneoNaGTvQZ:bfzhjjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e56b362a009fae06d3be8567f75fe36e39330d/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L492)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Consider `require.Contains` instead, to avoid the regexp escapes.
</blockquote></details>

Ah, missed your comment, thanks for the review Erik!
I think your suggestion makes sense, especially for the test in lease_test.go, but maybe we should keep it in case the spans change..


<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:
```diff
@@ -447,28 +448,30 @@ func (sr *txnSpanRefresher) maybeRefreshPreemptivelyLocked(
 	canRefreshTxn, refreshTxn := roachpb.PrepareTransactionForRefresh(ba.Txn, ba.Txn.WriteTimestamp)
 	if !canRefreshTxn || !sr.canAutoRetry {
-		return roachpb.BatchRequest{}, newRetryErrorOnFailedPreemptiveRefresh(ba.Txn)
+		return roachpb.BatchRequest{}, newRetryErrorOnFailedPreemptiveRefresh(ba.Txn, sr.refreshFootprint.asSlice())
```
I'm not confident that `asSlice` returns immutable memory. For example, look here:

https://github.com/cockroachdb/cockroach/blob/66903db87e4712a28bd2dc07424cfafecd91ae23/pkg/kv/kvclient/kvcoord/condensable_span_set.go#L50-L57

This seems to mutate the memory backing `s.s` in-place, so by referencing it from the error here and below, aren't we causing a data race?

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @lidorcarmel and @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MpzfjJ-EASsgAC2YW6Q:-MqJGfRl1qLj8S9Gosp-:b8s1l1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e56b362a009fae06d3be8567f75fe36e39330d/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L492)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

Ah, missed your comment, thanks for the review Erik!
I think your suggestion makes sense, especially for the test in lease_test.go, but maybe we should keep it in case the spans change..
</blockquote></details>

Ah, of course, I missed the actual regexp operators in there. :) Never mind then.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
It looks like this change propagates all of the refresh spans that have been accumulated by a transaction when a refresh fails. My reading of #61838 is that we wanted to begin propagating the specific read-write conflict that prevented the transaction from successfully refreshing. So we don't want to say "transaction retry because refresh failed after reading a, b, c, &d". Instead, we want to say "transaction retry because refresh observed read/write contention on c". Or something like that.

To do that, I think we're going to need to broaden the scope of this PR slightly. We already have this information on the client, because it is returned by the `Refresh` and `RefreshRange` commands. But the errors are unstructured ([here](https://github.com/cockroachdb/cockroach/blob/bad623e970b5199da3312d5b2cc611c26649d50c/pkg/kv/kvserver/batcheval/cmd_refresh.go#L66), [here](https://github.com/cockroachdb/cockroach/blob/bad623e970b5199da3312d5b2cc611c26649d50c/pkg/kv/kvserver/batcheval/cmd_refresh.go#L73), [here](https://github.com/cockroachdb/cockroach/blob/bad623e970b5199da3312d5b2cc611c26649d50c/pkg/kv/kvserver/batcheval/cmd_refresh_range.go#L68), and [here](https://github.com/cockroachdb/cockroach/blob/bad623e970b5199da3312d5b2cc611c26649d50c/pkg/kv/kvserver/batcheval/cmd_refresh_range.go#L84)).

I think we can break the issue into two parts. Let's first make this a structured error (see `pkg/roachpb/errors.proto`) so that we can pull out the information that we want on the client. Then let's use the structured error on the client to add more context to these retry errors that could not be refreshed.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqKaSJM69t-sUD-SBKo:b-tj0l3n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @lidorcarmel)



<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
First point, regarding printing the fail key only - definitely, this was not the right thing, thanks!
Second point, I did part 1 here, the structured error - is this what you had in mind? if yes then I can do the rest in a separate commit or pr.
Thanks!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqQKkfj1d61HZjQim6o:b-ch2763)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 451 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MpxSxyyWhEIy8ey5j-o-r2-451:-MqQJd6f6BywHCsltSHK:bd4dalt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed327450fe8a7206765f26c666e83070eaf113dd/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L451)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm not confident that `asSlice` returns immutable memory. For example, look here:

https://github.com/cockroachdb/cockroach/blob/66903db87e4712a28bd2dc07424cfafecd91ae23/pkg/kv/kvclient/kvcoord/condensable_span_set.go#L50-L57

This seems to mutate the memory backing `s.s` in-place, so by referencing it from the error here and below, aren't we causing a data race?
</blockquote></details>

and in addition this was not the right thing to use, removed all of that now.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I don't think we need to add errors to `errors.proto` and that errors union any more; I don't think we're adding errors there any more. Any error now makes it through RPCs if it knows how to serialize itself. Look for `RegisterLeafEncoder/RegisterWrapperEncoder`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqRDBIHBY8kL6KeMANy:bx3ht8a)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
Interesting. IIUC I need to implement the encode/decode myself instead of letting protos do this for me? mmm.. wondering about backward/forward compatibility issues, but maybe I'm missing something, I'll look better tomorrow. Thanks Andrei!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqSYFAuDa5INpYvCrO5:b-3uyy1z)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
You should be able to use any error type now. There are compatibility
issues in the sense that only new nodes will be able to understand the
error - others will treat it as opaque. This needs to be ok or we need a
version gate that makes sure the new error is emitted only once the cluster
has fully upgraded.

nvb reviewed (commented):
This new error type looks good to me, so I would proceed with this approach. But I haven't closely followed the adoption of the new errors library and the impact that has on proto encoded errors, so Tobi would know best.

@tbg do you have an example of its use that Lidor could look at? Specifically, a case where a new error type was added that didn't make its way into this `ErrorDetail` union. The only one I can find is `lockSpansOverBudgetError`, and that does not look like a good precedent to follow. We still want a protobuf-defined error type for cases like these, because forwards and backwards compatibility is important. [Code like this](https://github.com/cockroachdb/cockroach/blob/3d8e2dd2815772398841a566ef2d73110d3893df/pkg/kv/kvclient/kvcoord/lock_spans_over_budget_error.go#L68-L74) looks like a change in the wrong direction.

> Reviewed 2 of 9 files at r3, 4 of 6 files at r4, 4 of 4 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqWGEgK2cw8-aKJCpl9:bopi99x)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @lidorcarmel, and @tbg)

___
*[pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go, line 147 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFNqhDfIUtv7uIkh2:-MqWFNqhDfIUtv7uIkh3:b-1tzg9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go#L147)):*
> ```Go
> // TestRefreshRangeError verifies we get an error. We are trying to refresh from
> // time 1 to 3, but the key was written at time 2.
> func TestRefreshRangeError(t *testing.T) {
> ```

Consider merging these two tests using `testutils.RunTrueAndFalse(t, "intent", ...`. I think the only difference is whether you merge the intent or not.
___
*[pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go, line 179 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFGCdDEZ2EPFMqqWh:-MqWFGCdDEZ2EPFMqqWi:bh0hf15) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go#L179)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if err := db.Flush(); err != nil {
> ```

You shouldn't need to Flush or Compact. The `db` will still guarantee that you read your writes.
___
*[pkg/kv/kvserver/batcheval/cmd_refresh_test.go, line 29 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFb_4E2FF-YU3o9gL:-MqWFb_4E2FF-YU3o9gM:b-tb3ouy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/kv/kvserver/batcheval/cmd_refresh_test.go#L29)):*
> ```Go
> // TestRefreshError verifies we get an error. We are trying to refresh from
> // time 1 to 3, but the key was written at time 2.
> func TestRefreshError(t *testing.T) {
> ```

The same comments apply here.
___
*[pkg/roachpb/errors.go, line 1335 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFmdbDTDfQZkw3Noy:-MqWFmdc9Vry-wNy0nEo:b-k3i57g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/roachpb/errors.go#L1335)):*
> ```Go
> func (e *RefreshSpanError) message(_ *Error) string {
> 	keyOrIntent := "key"
> 	if e.Reason == RefreshSpanError_REASON_INTENT {
> ```

Consider using a switch-case with a default clause to handle forwards compatibility of a new reason.
___
*[pkg/roachpb/errors.proto, line 548 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFwTu6K6q0uHSbBDG:-MqWFwTu6K6q0uHSbBDH:bkv2ihl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/roachpb/errors.proto#L548)):*
> ```protobuf
> }
> 
> // A RefreshSpanError indicates that a refresh or a range refresh failed because
> ```

Consider naming this `RefreshFailedError` to make it clear that we're talking about a "failed refresh", as opposed to an error for some other reason.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I think we're indeed lacking good precedent. But, Lidor willing, I think we should take this opportunity to see if we can create one. Or, if we can't get something satisfying, to motivate some additions to cockroachdb/errors (which shouldn't block this PR; I don't want to hold this one up too long :P ).

It seems to me like we should be able to marry proto errors with the facilities in cockroachdb/errors pretty well. I think that, if you declare a new error type as a proto (because you want backwards and forward compatibility for it), you can fairly easily tell the library to use proto encoding/decoding when marshaling it.
The encoders that you specify for errors are alllowed to return arbitrary protos:
https://github.com/cockroachdb/errors/blob/985d51314b8d1cccb20a34336d5b08f019f0358c/errbase/encode.go#L284

So, if your error is already a proto, I think you can just return it. Something like:

```go
func encodeMyError(
	_ context.Context, err error,
) (msgPrefix string, safe []string, payload proto.Message) {
	t := err.(myErrorType)
	msgPrefix = "I don't know what this is"
	return msgPrefix, nil, t
}
```

Similarly, on the decoding side, the decoder takes in a proto `Message` and can type assert it to whatever it wants. I think it all works courtesy of the magic of the DynamicAny protobuf, with which the payload is unmarshalled. I might be dumb, though.

Lidor, would you be interested in seeing if this checks out? If not, as far as this PR and I is concerned, I think we can also merge it as it is.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqWPjxGDGySB_VKh2CJ:bb1qpqc)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @lidorcarmel, and @tbg)



<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
Thanks Andrei, I'm definitely willing :) I'll try that, maybe in another PR so that we can compare it to what we have here now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqXhp5N7VLVGLhsb85l:bcufqcy)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @lidorcarmel, and @tbg)



<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
cc @knz - FYI.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MqXkeNP79pRnQRJCZwa:b-h0dx6v)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @lidorcarmel, and @tbg)



<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
Addressed Nathan's comments, thanks!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-Mq_f5yX68VKCDM55saZ:bhnzmkf)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go, line 147 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFNqhDfIUtv7uIkh2:-Mq_YB6RCx_lvHBBpBjo:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go#L147)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider merging these two tests using `testutils.RunTrueAndFalse(t, "intent", ...`. I think the only difference is whether you merge the intent or not.
</blockquote></details>

Done, thanks!
___
*[pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go, line 179 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFGCdDEZ2EPFMqqWh:-MqXpTsPE4cDkW96DA74:bljmgoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/kv/kvserver/batcheval/cmd_refresh_range_test.go#L179)):*
> testutils.RunTrueAndFalse
Done.
___
*[pkg/kv/kvserver/batcheval/cmd_refresh_test.go, line 29 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFb_4E2FF-YU3o9gL:-Mq_dVEQBlJjaPLlsKHx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/kv/kvserver/batcheval/cmd_refresh_test.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The same comments apply here.
</blockquote></details>

Done.
___
*[pkg/roachpb/errors.go, line 1335 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFmdbDTDfQZkw3Noy:-Mq_ZGL1CymfFdcYrFHp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/roachpb/errors.go#L1335)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider using a switch-case with a default clause to handle forwards compatibility of a new reason.
</blockquote></details>

Done.
___
*[pkg/roachpb/errors.proto, line 548 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-MqWFwTu6K6q0uHSbBDG:-Mq_dOUN6zplqoD70N91:b-eeghma) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae35e853f719c7181299745590e96f20d1b2507/pkg/roachpb/errors.proto#L548)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider naming this `RefreshFailedError` to make it clear that we're talking about a "failed refresh", as opposed to an error for some other reason.
</blockquote></details>

Done, makes more sense, it's not a span as I originally thought..


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
Hi @andreimatei and others, see the draft PR #73827 for an alternative way to use the new style structured error. I know there are other ways to do this, mostly around where to put what, but I think overall it's not significantly better than the current PR, so I'd say let's continue with the current PR as is. WDYT?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-Mqv9S8R31ZmukVS2Woz:bcnn25b)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

Commit:
kvcoord: extend txn refresh error with the failed keys

Use the new structured error to log the actual key that we failed
refreshing. Instead of:
.. failed preemptive refresh ..

The new error will print:
.. failed preemptive refresh due to a conflict: committed value on key <key>
.. failed preemptive refresh due to a conflict: intent on key <key>

Rename tryUpdatingTxnSpans() to tryRefreshTxnSpans().

Fixes #61838.

Release note: None

nvb reviewed (commented):
> Reviewed 12 of 12 files at r8, 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @lidorcarmel and @tbg)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 314 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr991yLBXXjU2p6t29c:-Mr991yLBXXjU2p6t29d:bsdbzks) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L314)):*
> ```Go
> 
> 	// Try updating the txn spans so we can retry.
> 	if err := sr.tryUpdatingTxnSpans(ctx, refreshTxn); err != nil {
> ```

`s/err/refreshErr/`
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 316 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr98qwO8DWnCibnRICd:-Mr98qwO8DWnCibnRICe:b-ufzam5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L316)):*
> ```Go
> 	if err := sr.tryUpdatingTxnSpans(ctx, refreshTxn); err != nil {
> 		log.Eventf(ctx, "refresh failed; propagating original retry error")
> 		return nil, pErr
> ```

This was the second place where we return a retriable error in response to a failed transaction refresh (from our email thread). In this case, we have a retriable error that was generated by the server, and we're trying to "refresh it away". If we can't, we propagate the original error.

We're currently still not using the information from `err` to augment that error in the case where the refresh fails. I think that's ok for now, as we don't need to solve everything here and can continue to iterate on this in the future, but let's at least leave a TODO that references https://github.com/cockroachdb/cockroach/issues/41057.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 468 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr99uN5Bo0rPuMt2v43:-Mr99uN5Bo0rPuMt2v44:bsdbzks) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L468)):*
> ```Go
> 
> 	// Try updating the txn spans at a timestamp that will allow us to commit.
> 	if err := sr.tryUpdatingTxnSpans(ctx, refreshTxn); err != nil {
> ```

`s/err/refreshErr/`
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 479 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr99xlM9K2zExSbBS1E:-Mr99xlM9K2zExSbBS1F:b-t3po4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L479)):*
> ```Go
> 
> func newRetryErrorOnFailedPreemptiveRefresh(
> 	txn *roachpb.Transaction, pErr *roachpb.Error,
> ```

`s/pErr/refreshErr/`
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 488 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9A5j76jOtNTDj7ZY1:-Mr9A5j76jOtNTDj7ZY2:bqcfwll) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L488)):*
> ```Go
> 	if pErr != nil {
> 		if refreshErr, ok := pErr.GetDetail().(*roachpb.RefreshFailedError); ok {
> 			msg = fmt.Sprintf("%s of %s %s", msg, refreshErr.FailureReason(), refreshErr.Key)
> ```

Consider rewording this to something like `%s due to %s on key %s`.

EDIT: this suggestion doesn't work without addressing my other comment down below in `errors.go`.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 490 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9ADzj566E1368IwL6:-Mr9ADzj566E1368IwL7:bxu5wjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L490)):*
> ```Go
> 			msg = fmt.Sprintf("%s of %s %s", msg, refreshErr.FailureReason(), refreshErr.Key)
> 		} else {
> 			msg = fmt.Sprintf("%s - unknown error: %s", msg, pErr.String())
> ```

nit: there's no need to call `.String()`.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 505 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9ANXi9rRhWNfunBxw:-Mr9ANXi9rRhWNfunBxx:b8bxkms) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L505)):*
> ```Go
> // The provided transaction should be a Clone() of the original transaction with
> // its ReadTimestamp adjusted by the Refresh() method.
> func (sr *txnSpanRefresher) tryUpdatingTxnSpans(
> ```

Do you mind adding one more commit that renames this to `tryUpdatingTxnSpans` to `tryRefreshTxnSpans`? This naming hasn't aged well.
___
*[pkg/roachpb/errors.go, line 1337 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9ApsC5QxFEgcMutK4:-Mr9ApsC5QxFEgcMutK5:b-jvvezj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/roachpb/errors.go#L1337)):*
> ```Go
> 	var r string
> 	switch e.Reason {
> 	case RefreshFailedError_REASON_KEY:
> ```

On a second read, "key" doesn't really make sense here. Intents also have associated keys. I think what we actually want this case to represent is a conflict with a "committed value".
___
*[pkg/ts/catalog/chart_catalog.go, line 430 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9AdgO0ZyepFtas0-F:-Mr9AdgO0ZyepFtas0-G:b-96adxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/ts/catalog/chart_catalog.go#L430)):*
> ```Go
> 					"distsender.rpc.err.oprequirestxnerrtype",
> 					"distsender.rpc.err.optimisticevalconflictserrtype",
> 					"distsender.rpc.err.refreshfailederrtype",
> ```

I think this change was meant to be part of the previous commit.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
Thanks @nvanbenschoten, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MrA3XYl3B-Iq3V1caPy:b1y4fjh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 314 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr991yLBXXjU2p6t29c:-Mr9N3geCKfHSAUWyXA7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L314)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/err/refreshErr/`
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 316 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr98qwO8DWnCibnRICd:-Mr9Qr51BgUVouolfYlR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L316)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This was the second place where we return a retriable error in response to a failed transaction refresh (from our email thread). In this case, we have a retriable error that was generated by the server, and we're trying to "refresh it away". If we can't, we propagate the original error.

We're currently still not using the information from `err` to augment that error in the case where the refresh fails. I think that's ok for now, as we don't need to solve everything here and can continue to iterate on this in the future, but let's at least leave a TODO that references https://github.com/cockroachdb/cockroach/issues/41057.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 468 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr99uN5Bo0rPuMt2v43:-Mr9R-XCBbiL4CKSNAHS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L468)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/err/refreshErr/`
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 479 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr99xlM9K2zExSbBS1E:-Mr9JqOu8d6jOkhdBkjc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L479)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/pErr/refreshErr/`
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 488 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9A5j76jOtNTDj7ZY1:-Mr9RKKZ08ZCUH-lC-4D:b-4gjeai) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L488)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider rewording this to something like `%s due to %s on key %s`.

EDIT: this suggestion doesn't work without addressing my other comment down below in `errors.go`.
</blockquote></details>

Done as we said in slack.
FTR this would give us:
failed preemptive refresh due to a conflict: committed value on key abc
failed preemptive refresh due to a conflict: intent on key abc
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 490 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9ADzj566E1368IwL6:-Mr9JxruA1uF5AVJ0p11:b8ii3de) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L490)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: there's no need to call `.String()`.
</blockquote></details>

of course not..
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 505 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9ANXi9rRhWNfunBxw:-MrA1JVp0R9N3YFzXybl:b-8aikcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L505)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you mind adding one more commit that renames this to `tryUpdatingTxnSpans` to `tryRefreshTxnSpans`? This naming hasn't aged well.
</blockquote></details>

Of course. Done.
___
*[pkg/roachpb/errors.go, line 1337 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9ApsC5QxFEgcMutK4:-Mr9LFSfEP8h615tBp3H:b2ybdv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/roachpb/errors.go#L1337)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

On a second read, "key" doesn't really make sense here. Intents also have associated keys. I think what we actually want this case to represent is a conflict with a "committed value".
</blockquote></details>

Makes more sense, thanks.
___
*[pkg/ts/catalog/chart_catalog.go, line 430 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-Mr9AdgO0ZyepFtas0-F:-Mr9yu5_AvhZ9FCsWfFD:bcq4d07) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed189352e11393556123aab181d6a4e67d5f36d5/pkg/ts/catalog/chart_catalog.go#L430)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think this change was meant to be part of the previous commit.
</blockquote></details>

I think you're right, done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 8 of 13 files at r10, 5 of 5 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73413#-:-MrAKuXu-0BFVsqMY2i7:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)



<!-- Sent from Reviewable.io -->

lidorcarmel commented:
bors r+
